test/unit/asrouter/CFRPageActions.test.js

```diff
@@ -0,0 +1,641 @@
+      it("should call ._dispatchToASRouter with the right action", () => {
+        const fakeAction = {};
+        pageAction.dispatchUserAction(fakeAction);
+        assert.calledOnce(dispatchStub);
+        assert.calledWith(
+          dispatchStub,
+          {type: "USER_ACTION", data: fakeAction},
+          {browser: fakeBrowser}
+        );
+      });
+    });
+
+    describe("#_dispatchImpression", () => {
+      it("should call ._dispatchToASRouter with the right action", () => {
+        pageAction._dispatchImpression("fake impression");
+        assert.calledWith(dispatchStub, {type: "IMPRESSION", data: "fake impression"});
+      });
+    });
+
+    describe("#_sendTelemetry", () => {
+      it("should call ._dispatchToASRouter with the right action", () => {
+        const fakePing = {message_id: 42};
+        pageAction._sendTelemetry(fakePing);
+        assert.calledWith(dispatchStub, {
+          type: "DOORHANGER_TELEMETRY",
+          data: {useClientID: true, action: "cfr_user_event", source: "CFR", message_id: 42}
+        });
+      });
+    });
+
+    describe("#_blockMessage", () => {
+      it("should call ._dispatchToASRouter with the right action", () => {
+        pageAction._blockMessage("fake id");
+        assert.calledOnce(dispatchStub);
+        assert.calledWith(dispatchStub, {
+          type: "BLOCK_MESSAGE_BY_ID",
+          data: {id: "fake id"}
+        });
+      });
+    });
+
+    describe("#getStrings", () => {
+      const localeStrings = [{
+        value: "你好世界",
+        attributes: [
+          {name: "first_attr", value: 42},
+          {name: "second_attr", value: "some string"},
+          {name: "third_attr", value: [1, 2, 3]}
+        ]
+      }];
+
+      beforeEach(() => {
+        getStringsStub.restore();
+        global.Localization.prototype.formatMessages = sandbox.stub()
+          .withArgs({id: "hello_world"})
+          .resolves(localeStrings);
+      });
+
+      it("should return the argument if a string_id is not defined", async () => {
+        assert.deepEqual(await pageAction.getStrings({}), {});
+        assert.equal(await pageAction.getStrings("some string"), "some string");
+      });
+      it("should get the right locale string", async () => {
+        assert.equal(await pageAction.getStrings({string_id: "hello_world"}), localeStrings[0].value);
+      });
+      it("should return the right sub-attribute if specified", async () => {
+        assert.equal(await pageAction.getStrings({string_id: "hello_world"}, "second_attr"), "some string");
+      });
+    });
+
+    describe("#_handleClick", () => {
+      beforeEach(async () => {
+        CFRPageActions.PageActionMap.set(fakeBrowser.ownerGlobal, pageAction);
+        await CFRPageActions.addRecommendation(fakeBrowser, fakeHost, fakeRecommendation, dispatchStub);
+        // Reset the call history for all spies and stubs so that we needn't consider the effects of the above
+        sandbox.resetHistory();
```

Comment:
Is there any way we can work around this? It seems like a weird hack, at most I would isolate this call to the specific tests (`it` blocks) that require it otherwise it's easy to overlook when adding new tests.

Reply:
You're right, it is a bit weird, luckily it seems to not actually be necessary so I've removed it